[New post] Get Started with Spring Boot, SAML, and Okta (#1233)
oktadeveloper/okta-blog: _source/_posts/2022-08-09-spring-boot-saml.adoc

My recommendation: just use OIDC.

If you _must_ use SAML with Spring Boot, this tutorial should make it quick and easy.

**Prerequisites:**

- https://sdkman.io/[SDKMAN] (for Java 17)

toc::[]

== What is SAML?

**S**ecurity **A**ssertion **M**arkup **L**anguage is an XML-based way of doing web authentication and authorization. It works cross-domain, so SaaS applications and other enterprise software often support it.

Nick Gamb has an excellent overview in https://www.youtube.com/watch?v=l-6QSEqDJPo[A Developer's Guide to SAML].

++++
{% youtube l-6QSEqDJPo %}
++++

See Spring Security's documentation for a https://docs.spring.io/spring-security/reference/servlet/saml2/login/overview.html[SAML 2.0 Login Overview].

== Add a SAML application on Okta

To begin, you'll need an Okta developer account. You can create one at https://developer.okta.com/signup[developer.okta.com/signup] or install the https://cli.okta.com[Okta CLI] and run `okta register`.

Then, log in to your account and go to *Applications* > *Create App Integration*. Select *SAML 2.0* and click *Next*. Name your app something like `Spring Boot SAML` and click *Next*.

Use the following settings:

> * Single sign on URL: `\http://localhost:8080/login/saml2/sso/okta`

Reviewer: ```suggestion
* Single sign-on URL: `\http://localhost:8080/login/saml2/sso/okta`
```
nit

Author: Unfortunately, this is how it's spelled in the Okta Admin Console. 😞

<img width="707" alt="Screen Shot 2022-08-04 at 13 17 52" src="https://user-images.githubusercontent.com/17892/182934966-2c4dd5fc-4379-443b-803f-aaf3c2ebc5f9.png">

  }

  const rolesClaim = prepareCustomClaimKey('roles');

  if (context.idToken) {
    context.idToken[rolesClaim] = roles;
  }

  if (context.accessToken) {
    context.accessToken[rolesClaim] = roles;
  }

  callback(null, user, context);
}
----

Next, modify `HomeController` to allow Auth0's email attribute name.

[source,java]
----
public class HomeController {

    @RequestMapping("/")
    public String home(@AuthenticationPrincipal Saml2AuthenticatedPrincipal principal, Model model) {
        model.addAttribute("name", principal.getName());
        String email = principal.getFirstAttribute("email");
        // if email is not preset, try Auth0 attribute name
        if (email == null) {
            email = principal.getFirstAttribute("http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress");
        }
        model.addAttribute("emailAddress", email);
        model.addAttribute("userAttributes", principal.getAttributes());
        return "home";
    }

}
----

Restart your app, log in, and everything should work as expected.

image::{% asset_path 'blog/spring-boot-saml2/auth0-groups.png' %}[alt=Auth0 groups,width=800,align=center]

== Support Okta and Auth0

> You can also support _both_ Okta and Auth0! Modify your `application.yml` to be as follows, and you'll be prompted by Spring Security for which one to log in with. The `&name` and `*name` values are used to set and retrieve blocks of YAML to avoid repetition.

Reviewer: ```suggestion
You can also support _both_ Okta and Auth0! Modify your `application.yml` to be as follows, and Spring Security will prompt you for which one to log in with. The `&name` and `*name` values are used to set and retrieve blocks of YAML to avoid repetition.
```
changes passive voice to active
